File: backend/data_processing/data_cleaner.js
const fs = require('fs');
const csv = require('csv-parser');
const moment = require('moment');
const path = require('path');

/**
 * Data Cleaner for NYC Taxi Trip Dataset
 * Handles missing values, duplicates, invalid records, and outliers
 * Calculates derived features as required by assignment
 */
class DataCleaner {
    constructor() {
        this.excludedRecords = [];
        this.suspiciousRecords = [];
        this.processedRecords = [];
        
        // NYC bounds for coordinate validation
        this.nycBounds = {
            minLat: 40.4774,  // Staten Island
            maxLat: 40.9176,  // Bronx
            minLon: -74.2591, // New Jersey border
            maxLon: -73.7004  // Queens
        };
    }

    /**
     * Main data cleaning pipeline
     * @param {string} inputFile - Path to input CSV file
     * @param {string} outputFile - Path to output cleaned CSV file
     */
    async cleanData(inputFile, outputFile) {
        console.log('🧹 Starting data cleaning pipeline...');
        console.log(`📁 Input: ${inputFile}`);
        console.log(`📁 Output: ${outputFile}`);

        const rawData = [];
        
        // Read raw data
        await new Promise((resolve, reject) => {
            fs.createReadStream(inputFile)
                .pipe(csv())
                .on('data', (row) => rawData.push(row))
                .on('end', resolve)
                .on('error', reject);
        });

        console.log(`📊 Loaded ${rawData.length} raw records`);

        let processedCount = 0;
        let excludedCount = 0;

        // Process each record
        for (let i = 0; i < rawData.length; i++) {
            const record = rawData[i];
            
            try {
                const cleanedRecord = this.cleanRecord(record, i);
                
                if (cleanedRecord) {
                    this.processedRecords.push(cleanedRecord);
                    processedCount++;
                } else {
                    excludedCount++;
                }
            } catch (error) {
                console.warn(`⚠️  Error processing record ${i}: ${error.message}`);
                this.excludedRecords.push({
                    index: i,
                    reason: 'processing_error',
                    error: error.message
                });
                excludedCount++;
            }

            // Progress indicator
            if (i % 10000 === 0) {
                console.log(`🔄 Processed ${i}/${rawData.length} records...`);
            }
        }

        // Write cleaned data
        await this.writeCleanedData(outputFile);

        // Generate summary report
        this.generateSummaryReport(rawData.length, processedCount, excludedCount);

        return {
            totalRecords: rawData.length,
            processedRecords: processedCount,
            excludedRecords: excludedCount,
            exclusionRate: (excludedCount / rawData.length * 100).toFixed(2)
        };
    }

    /**
     * Clean individual record
     * @param {Object} record - Raw record from CSV
     * @param {number} index - Record index
     * @returns {Object|null} - Cleaned record or null if excluded
     */
    cleanRecord(record, index) {
        // 1. Handle missing values
        if (!this.handleMissingValues(record)) {
            this.excludedRecords.push({
                index,
                reason: 'missing_critical_data',
                details: 'Missing pickup/dropoff datetime or coordinates'
            });
            return null;
        }

        // 2. Validate and clean timestamps
        if (!this.validateTimestamps(record)) {
            this.excludedRecords.push({
                index,
                reason: 'invalid_timestamps',
                details: 'Invalid or inconsistent pickup/dropoff times'
            });
            return null;
        }

        // 3. Validate coordinates
        if (!this.validateCoordinates(record)) {
            this.excludedRecords.push({
                index,
                reason: 'invalid_coordinates',
                details: 'Coordinates outside NYC bounds or invalid'
            });
            return null;
        }

        // 4. Validate trip metrics
        if (!this.validateTripMetrics(record)) {
            this.excludedRecords.push({
                index,
                reason: 'invalid_trip_metrics',
                details: 'Invalid passenger count, distance, or duration'
            });
            return null;
        }

        // 5. Validate fare data
        if (!this.validateFareData(record)) {
            this.excludedRecords.push({
                index,
                reason: 'invalid_fare_data',
                details: 'Invalid fare amounts or payment type'
            });
            return null;
        }

        // 6. Calculate derived features (Assignment Requirement)
        this.calculateDerivedFeatures(record);

        // 7. Check for outliers
        if (this.isOutlier(record)) {
            this.excludedRecords.push({
                index,
                reason: 'outlier',
                details: 'Record exceeds reasonable bounds'
            });
            return null;
        }

        return record;
    }

    /**
     * Handle missing values
     * @param {Object} record - Record to clean
     * @returns {boolean} - True if record is valid after cleaning
     */
    handleMissingValues(record) {
        // Critical fields that cannot be missing
        const criticalFields = [
            'pickup_datetime',
            'dropoff_datetime', 
            'pickup_longitude',
            'pickup_latitude',
            'dropoff_longitude',
            'dropoff_latitude'
        ];

        for (const field of criticalFields) {
            if (!record[field] || record[field].trim() === '') {
                return false;
            }
        }

        // Fill missing values for other fields
        record.passenger_count = record.passenger_count || 1;
        record.trip_distance = record.trip_distance || 0;
        record.fare_amount = record.fare_amount || 0;
        record.tip_amount = record.tip_amount || 0;
        record.tolls_amount = record.tolls_amount || 0;
        record.total_amount = record.total_amount || 0;
        record.payment_type = record.payment_type || 1;

        return true;
    }

    /**
     * Validate timestamps
     * @param {Object} record - Record to validate
     * @returns {boolean} - True if timestamps are valid
     */
    validateTimestamps(record) {
        try {
            const pickupTime = moment(record.pickup_datetime);
            const dropoffTime = moment(record.dropoff_datetime);

            // Check if dates are valid
            if (!pickupTime.isValid() || !dropoffTime.isValid()) {
                return false;
            }

            // Check if dropoff is after pickup
            if (dropoffTime.isBefore(pickupTime)) {
                return false;
            }

            // Check if dates are within reasonable range (2015-2020)
            const startDate = moment('2015-01-01');
            const endDate = moment('2020-12-31');
            
            if (pickupTime.isBefore(startDate) || pickupTime.isAfter(endDate)) {
                return false;
            }

            // Calculate trip duration
            record.trip_duration = dropoffTime.diff(pickupTime, 'seconds');

            // Check if duration is reasonable (30 seconds to 24 hours)
            if (record.trip_duration < 30 || record.trip_duration > 86400) {
                return false;
            }

            return true;
        } catch (error) {
            return false;
        }
    }

    /**
     * Validate coordinates
     * @param {Object} record - Record to validate
     * @returns {boolean} - True if coordinates are valid
     */
    validateCoordinates(record) {
        const coords = [
            'pickup_longitude',
            'pickup_latitude',
            'dropoff_longitude',
            'dropoff_latitude'
        ];

        for (const coord of coords) {
            const value = parseFloat(record[coord]);
            if (isNaN(value)) {
                return false;
            }
            record[coord] = value;
        }

        // Check if coordinates are within NYC bounds
        if (record.pickup_latitude < this.nycBounds.minLat || 
            record.pickup_latitude > this.nycBounds.maxLat ||
            record.pickup_longitude < this.nycBounds.minLon || 
            record.pickup_longitude > this.nycBounds.maxLon ||
            record.dropoff_latitude < this.nycBounds.minLat || 
            record.dropoff_latitude > this.nycBounds.maxLat ||
            record.dropoff_longitude < this.nycBounds.minLon || 
            record.dropoff_longitude > this.nycBounds.maxLon) {
            return false;
        }

        return true;
    }

    /**
     * Validate trip metrics
     * @param {Object} record - Record to validate
     * @returns {boolean} - True if metrics are valid
     */
    validateTripMetrics(record) {
        // Passenger count (1-6)
        record.passenger_count = parseInt(record.passenger_count);
        if (record.passenger_count < 1 || record.passenger_count > 6) {
            return false;
        }

        // Trip distance (0-500 miles)
        record.trip_distance = parseFloat(record.trip_distance);
        if (record.trip_distance < 0 || record.trip_distance > 500) {
            return false;
        }

        return true;
    }

    /**
     * Validate fare data
     * @param {Object} record - Record to validate
     * @returns {boolean} - True if fare data is valid
     */
    validateFareData(record) {
        // Fare amounts
        record.fare_amount = parseFloat(record.fare_amount);
        record.tip_amount = parseFloat(record.tip_amount || 0);
        record.tolls_amount = parseFloat(record.tolls_amount || 0);
        record.total_amount = parseFloat(record.total_amount);

        // Check fare amounts are reasonable
        if (record.fare_amount < 0 || record.fare_amount > 1000 ||
            record.tip_amount < 0 || record.tip_amount > 500 ||
            record.tolls_amount < 0 || record.tolls_amount > 100 ||
            record.total_amount < 0 || record.total_amount > 2000) {
            return false;
        }

        // Payment type
        record.payment_type = parseInt(record.payment_type);
        if (record.payment_type < 1 || record.payment_type > 6) {
            return false;
        }

        return true;
    }

    /**
     * Calculate derived features (Assignment Requirement #3)
     * @param {Object} record - Record to enhance
     */
    calculateDerivedFeatures(record) {
        // DERIVED FEATURE 1: Trip Speed (mph)
        // Justification: Identifies traffic patterns, route efficiency, and congestion
        if (record.trip_distance > 0 && record.trip_duration > 0) {
            const durationHours = record.trip_duration / 3600;
            record.trip_speed_mph = record.trip_distance / durationHours;
            
            // Cap unrealistic speeds
            if (record.trip_speed_mph > 100) {
                record.trip_speed_mph = 100;
            }
        } else {
            record.trip_speed_mph = 0;
        }

        // DERIVED FEATURE 2: Fare per Mile ($/mile)
        // Justification: Measures pricing efficiency, identifies premium routes
        if (record.trip_distance > 0) {
            record.fare_per_mile = record.fare_amount / record.trip_distance;
        } else {
            record.fare_per_mile = 0;
        }

        // DERIVED FEATURE 3: Idle Time (minutes)
        // Justification: Estimates time spent in traffic vs moving, congestion indicator
        if (record.trip_distance > 0 && record.trip_duration > 0) {
            // Assume average city speed is 12 mph
            const expectedMovingTime = (record.trip_distance / 12) * 60; // minutes
            const actualTime = record.trip_duration / 60; // minutes
            record.idle_time_minutes = Math.max(0, actualTime - expectedMovingTime);
        } else {
            record.idle_time_minutes = 0;
        }

        // Additional derived feature: Tip Percentage
        if (record.fare_amount > 0) {
            record.tip_percentage = (record.tip_amount / record.fare_amount) * 100;
        } else {
            record.tip_percentage = 0;
        }
    }

    /**
     * Check if record is an outlier
     * @param {Object} record - Record to check
     * @returns {boolean} - True if record is an outlier
     */
    isOutlier(record) {
        // Speed outliers
        if (record.trip_speed_mph > 80) {
            return true;
        }

        // Fare per mile outliers
        if (record.fare_per_mile > 50) {
            return true;
        }

        // Idle time outliers (more than 2 hours of idle time)
        if (record.idle_time_minutes > 120) {
            return true;
        }

        return false;
    }

    /**
     * Write cleaned data to CSV
     * @param {string} outputFile - Output file path
     */
    async writeCleanedData(outputFile) {
        console.log('💾 Writing cleaned data...');
        
        const headers = [
            'pickup_datetime',
            'dropoff_datetime',
            'passenger_count',
            'trip_distance',
            'trip_duration',
            'fare_amount',
            'tip_amount',
            'tolls_amount',
            'total_amount',
            'payment_type',
            'pickup_longitude',
            'pickup_latitude',
            'dropoff_longitude',
            'dropoff_latitude',
            'trip_speed_mph',
            'fare_per_mile',
            'idle_time_minutes',
            'tip_percentage'
        ];

        const csvContent = [
            headers.join(','),
            ...this.processedRecords.map(record => 
                headers.map(header => record[header] || '').join(',')
            )
        ].join('\n');

        fs.writeFileSync(outputFile, csvContent);
        console.log(`✅ Cleaned data written to: ${outputFile}`);
    }

    /**
     * Generate summary report
     * @param {number} totalRecords - Total input records
     * @param {number} processedRecords - Successfully processed records
     * @param {number} excludedRecords - Excluded records
     */
    generateSummaryReport(totalRecords, processedRecords, excludedRecords) {
        const report = {
            timestamp: new Date().toISOString(),
            summary: {
                totalRecords,
                processedRecords,
                excludedRecords,
                exclusionRate: (excludedRecords / totalRecords * 100).toFixed(2) + '%'
            },
            exclusionReasons: this.getExclusionSummary(),
            derivedFeatures: {
                trip_speed_mph: 'Calculated as distance/duration in mph - identifies traffic patterns',
                fare_per_mile: 'Calculated as fare_amount/distance - measures pricing efficiency',
                idle_time_minutes: 'Estimated time in traffic vs moving - congestion indicator'
            },
            dataQualityMetrics: this.calculateQualityMetrics()
        };

        const reportFile = path.join(path.dirname(outputFile), 'cleaning_report.json');
        fs.writeFileSync(reportFile, JSON.stringify(report, null, 2));
        
        console.log('\n📊 DATA CLEANING SUMMARY');
        console.log('========================');
        console.log(`Total Records: ${totalRecords.toLocaleString()}`);
        console.log(`Processed: ${processedRecords.toLocaleString()}`);
        console.log(`Excluded: ${excludedRecords.toLocaleString()}`);
        console.log(`Exclusion Rate: ${report.summary.exclusionRate}`);
        console.log(`📄 Detailed report: ${reportFile}`);
    }

    /**
     * Get summary of exclusion reasons
     * @returns {Object} - Summary of exclusions
     */
    getExclusionSummary() {
        const summary = {};
        this.excludedRecords.forEach(record => {
            summary[record.reason] = (summary[record.reason] || 0) + 1;
        });
        return summary;
    }

    /**
     * Calculate data quality metrics
     * @returns {Object} - Quality metrics
     */
    calculateQualityMetrics() {
        if (this.processedRecords.length === 0) {
            return {};
        }

        const speeds = this.processedRecords.map(r => r.trip_speed_mph).filter(s => s > 0);
        const faresPerMile = this.processedRecords.map(r => r.fare_per_mile).filter(f => f > 0);
        const idleTimes = this.processedRecords.map(r => r.idle_time_minutes).filter(i => i > 0);

        return {
            averageSpeed: speeds.length > 0 ? (speeds.reduce((a, b) => a + b, 0) / speeds.length).toFixed(2) + ' mph' : 'N/A',
            averageFarePerMile: faresPerMile.length > 0 ? '$' + (faresPerMile.reduce((a, b) => a + b, 0) / faresPerMile.length).toFixed(2) : 'N/A',
            averageIdleTime: idleTimes.length > 0 ? (idleTimes.reduce((a, b) => a + b, 0) / idleTimes.length).toFixed(2) + ' minutes' : 'N/A'
        };
    }
}

module.exports = DataCleaner;

File: backend/data_processing/data_cleaner.py
import pandas as pd
import numpy as np
from datetime import datetime, timedelta
import logging
import os
from typing import Tuple, Dict, List

# Set up logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class DataCleaner:
    """Handles data cleaning and preprocessing for NYC taxi data"""
    
    def __init__(self):
        self.excluded_records = []
        self.suspicious_records = []
        
    def load_data(self, file_path: str) -> pd.DataFrame:
        """Load raw CSV data"""
        try:
            logger.info(f"Loading data from {file_path}")
            
            # Define column names based on NYC taxi dataset format
            columns = [
                'vendor_id', 'pickup_datetime', 'dropoff_datetime',
                'passenger_count', 'trip_distance', 'pickup_longitude',
                'pickup_latitude', 'rate_code_id', 'store_and_fwd_flag',
                'dropoff_longitude', 'dropoff_latitude', 'payment_type',
                'fare_amount', 'extra', 'mta_tax', 'tip_amount',
                'tolls_amount', 'improvement_surcharge', 'total_amount'
            ]
            
            df = pd.read_csv(file_path, names=columns, low_memory=False)
            logger.info(f"Loaded {len(df)} records")
            return df
            
        except Exception as e:
            logger.error(f"Error loading data: {e}")
            raise
    
    def clean_data(self, df: pd.DataFrame) -> pd.DataFrame:
        """Main data cleaning pipeline"""
        logger.info("Starting data cleaning process")
        
        initial_count = len(df)
        
        # 1. Handle missing values
        df = self._handle_missing_values(df)
        
        # 2. Remove duplicates
        df = self._remove_duplicates(df)
        
        # 3. Validate and clean timestamps
        df = self._clean_timestamps(df)
        
        # 4. Validate and clean coordinates
        df = self._clean_coordinates(df)
        
        # 5. Validate trip metrics
        df = self._validate_trip_metrics(df)
        
        # 6. Validate fare data
        df = self._validate_fare_data(df)
        
        # 7. Calculate derived features
        df = self._calculate_derived_features(df)
        
        # 8. Remove outliers
        df = self._remove_outliers(df)
        
        final_count = len(df)
        excluded_count = initial_count - final_count
        
        logger.info(f"Data cleaning complete. Excluded {excluded_count} records ({excluded_count/initial_count*100:.2f}%)")
        logger.info(f"Final dataset: {final_count} records")
        
        return df
    
    def _handle_missing_values(self, df: pd.DataFrame) -> pd.DataFrame:
        """Handle missing values in the dataset"""
        logger.info("Handling missing values")
        
        # Remove rows with missing critical data
        critical_columns = ['pickup_datetime', 'dropoff_datetime', 'pickup_longitude', 
                          'pickup_latitude', 'dropoff_longitude', 'dropoff_latitude']
        
        missing_critical = df[critical_columns].isnull().any(axis=1)
        excluded_count = missing_critical.sum()
        
        if excluded_count > 0:
            self.excluded_records.extend([
                {'reason': 'missing_critical_data', 'count': excluded_count}
            ])
            logger.info(f"Excluding {excluded_count} records with missing critical data")
        
        df = df.dropna(subset=critical_columns)
        
        # Fill missing values for other columns
        df['passenger_count'] = df['passenger_count'].fillna(1)
        df['trip_distance'] = df['trip_distance'].fillna(0)
        df['fare_amount'] = df['fare_amount'].fillna(0)
        df['tip_amount'] = df['tip_amount'].fillna(0)
        df['tolls_amount'] = df['tolls_amount'].fillna(0)
        df['total_amount'] = df['total_amount'].fillna(0)
        df['payment_type'] = df['payment_type'].fillna(1)  # Default to credit card
        
        return df
    
    def _remove_duplicates(self, df: pd.DataFrame) -> pd.DataFrame:
        """Remove duplicate records"""
        logger.info("Removing duplicate records")
        
        initial_count = len(df)
        
        # Remove exact duplicates
        df = df.drop_duplicates()
        
        # Remove trips with same pickup/dropoff time and location (likely duplicates)
        df = df.drop_duplicates(subset=['pickup_datetime', 'dropoff_datetime', 
                                       'pickup_longitude', 'pickup_latitude',
                                       'dropoff_longitude', 'dropoff_latitude'], keep='first')
        
        duplicate_count = initial_count - len(df)
        
        if duplicate_count > 0:
            self.excluded_records.append({
                'reason': 'duplicate_records', 
                'count': duplicate_count
            })
            logger.info(f"Removed {duplicate_count} duplicate records")
        
        return df
    
    def _clean_timestamps(self, df: pd.DataFrame) -> pd.DataFrame:
        """Clean and validate timestamp data"""
        logger.info("Cleaning timestamps")
        
        # Convert to datetime
        df['pickup_datetime'] = pd.to_datetime(df['pickup_datetime'], errors='coerce')
        df['dropoff_datetime'] = pd.to_datetime(df['dropoff_datetime'], errors='coerce')
        
        # Remove invalid timestamps
        invalid_timestamps = df['pickup_datetime'].isnull() | df['dropoff_datetime'].isnull()
        invalid_count = invalid_timestamps.sum()
        
        if invalid_count > 0:
            self.excluded_records.append({
                'reason': 'invalid_timestamps',
                'count': invalid_count
            })
            logger.info(f"Excluding {invalid_count} records with invalid timestamps")
        
        df = df.dropna(subset=['pickup_datetime', 'dropoff_datetime'])
        
        # Remove trips where dropoff is before pickup
        invalid_order = df['dropoff_datetime'] < df['pickup_datetime']
        invalid_order_count = invalid_order.sum()
        
        if invalid_order_count > 0:
            self.excluded_records.append({
                'reason': 'dropoff_before_pickup',
                'count': invalid_order_count
            })
            logger.info(f"Excluding {invalid_order_count} records where dropoff is before pickup")
        
        df = df[df['dropoff_datetime'] >= df['pickup_datetime']]
        
        # Remove trips outside reasonable date range (2015-2020 for this dataset)
        start_date = pd.to_datetime('2015-01-01')
        end_date = pd.to_datetime('2020-12-31')
        
        date_out_of_range = (df['pickup_datetime'] < start_date) | (df['pickup_datetime'] > end_date)
        date_out_count = date_out_of_range.sum()
        
        if date_out_count > 0:
            self.excluded_records.append({
                'reason': 'date_out_of_range',
                'count': date_out_count
            })
            logger.info(f"Excluding {date_out_count} records outside date range")
        
        df = df[~date_out_of_range]
        
        return df
    
    def _clean_coordinates(self, df: pd.DataFrame) -> pd.DataFrame:
        """Clean and validate coordinate data"""
        logger.info("Cleaning coordinates")
        
        # Convert to numeric
        coord_columns = ['pickup_longitude', 'pickup_latitude', 'dropoff_longitude', 'dropoff_latitude']
        for col in coord_columns:
            df[col] = pd.to_numeric(df[col], errors='coerce')
        
        # Remove records with invalid coordinates
        invalid_coords = df[coord_columns].isnull().any(axis=1)
        invalid_coord_count = invalid_coords.sum()
        
        if invalid_coord_count > 0:
            self.excluded_records.append({
                'reason': 'invalid_coordinates',
                'count': invalid_coord_count
            })
            logger.info(f"Excluding {invalid_coord_count} records with invalid coordinates")
        
        df = df.dropna(subset=coord_columns)
        
        # Remove coordinates outside NYC area (rough bounds)
        nyc_bounds = {
            'min_lat': 40.4774,  # Staten Island
            'max_lat': 40.9176,  # Bronx
            'min_lon': -74.2591, # New Jersey border
            'max_lon': -73.7004  # Queens
        }
        
        outside_nyc = (
            (df['pickup_latitude'] < nyc_bounds['min_lat']) | 
            (df['pickup_latitude'] > nyc_bounds['max_lat']) |
            (df['pickup_longitude'] < nyc_bounds['min_lon']) | 
            (df['pickup_longitude'] > nyc_bounds['max_lon']) |
            (df['dropoff_latitude'] < nyc_bounds['min_lat']) | 
            (df['dropoff_latitude'] > nyc_bounds['max_lat']) |
            (df['dropoff_longitude'] < nyc_bounds['min_lon']) | 
            (df['dropoff_longitude'] > nyc_bounds['max_lon'])
        )
        
        outside_count = outside_nyc.sum()
        
        if outside_count > 0:
            self.excluded_records.append({
                'reason': 'coordinates_outside_nyc',
                'count': outside_count
            })
            logger.info(f"Excluding {outside_count} records with coordinates outside NYC")
        
        df = df[~outside_nyc]
        
        return df
    
    def _validate_trip_metrics(self, df: pd.DataFrame) -> pd.DataFrame:
        """Validate trip distance, duration, and passenger count"""
        logger.info("Validating trip metrics")
        
        # Validate passenger count (1-6 passengers)
        invalid_passengers = (df['passenger_count'] < 1) | (df['passenger_count'] > 6)
        invalid_passenger_count = invalid_passengers.sum()
        
        if invalid_passenger_count > 0:
            self.excluded_records.append({
                'reason': 'invalid_passenger_count',
                'count': invalid_passenger_count
            })
            logger.info(f"Excluding {invalid_passenger_count} records with invalid passenger count")
        
        df = df[~invalid_passengers]
        
        # Validate trip distance (0-500 miles)
        invalid_distance = (df['trip_distance'] < 0) | (df['trip_distance'] > 500)
        invalid_distance_count = invalid_distance.sum()
        
        if invalid_distance_count > 0:
            self.excluded_records.append({
                'reason': 'invalid_trip_distance',
                'count': invalid_distance_count
            })
            logger.info(f"Excluding {invalid_distance_count} records with invalid trip distance")
        
        df = df[~invalid_distance]
        
        # Flag suspicious records for analysis
        suspicious_distance = (df['trip_distance'] == 0) & (df['fare_amount'] > 0)
        suspicious_count = suspicious_distance.sum()
        
        if suspicious_count > 0:
            self.suspicious_records.append({
                'reason': 'zero_distance_nonzero_fare',
                'count': suspicious_count
            })
            logger.info(f"Flagged {suspicious_count} suspicious records with zero distance but nonzero fare")
        
        return df
    
    def _validate_fare_data(self, df: pd.DataFrame) -> pd.DataFrame:
        """Validate fare and payment data"""
        logger.info("Validating fare data")
        
        # Validate fare amount (0-1000)
        invalid_fare = (df['fare_amount'] < 0) | (df['fare_amount'] > 1000)
        invalid_fare_count = invalid_fare.sum()
        
        if invalid_fare_count > 0:
            self.excluded_records.append({
                'reason': 'invalid_fare_amount',
                'count': invalid_fare_count
            })
            logger.info(f"Excluding {invalid_fare_count} records with invalid fare amount")
        
        df = df[~invalid_fare]
        
        # Validate tip amount (0-500)
        invalid_tip = (df['tip_amount'] < 0) | (df['tip_amount'] > 500)
        invalid_tip_count = invalid_tip.sum()
        
        if invalid_tip_count > 0:
            self.excluded_records.append({
                'reason': 'invalid_tip_amount',
                'count': invalid_tip_count
            })
            logger.info(f"Excluding {invalid_tip_count} records with invalid tip amount")
        
        df = df[~invalid_tip]
        
        # Validate total amount
        invalid_total = (df['total_amount'] < 0) | (df['total_amount'] > 2000)
        invalid_total_count = invalid_total.sum()
        
        if invalid_total_count > 0:
            self.excluded_records.append({
                'reason': 'invalid_total_amount',
                'count': invalid_total_count
            })
            logger.info(f"Excluding {invalid_total_count} records with invalid total amount")
        
        df = df[~invalid_total]
        
        return df
    
    def _calculate_derived_features(self, df: pd.DataFrame) -> pd.DataFrame:
        """Calculate derived features"""
        logger.info("Calculating derived features")
        
        # Calculate trip duration in seconds
        df['trip_duration'] = (df['dropoff_datetime'] - df['pickup_datetime']).dt.total_seconds()
        
        # Calculate trip speed (mph) - Feature 1
        df['trip_speed_mph'] = np.where(
            df['trip_duration'] > 0,
            (df['trip_distance'] / (df['trip_duration'] / 3600)).replace([np.inf, -np.inf], np.nan),
            np.nan
        )
        
        # Calculate fare per mile - Feature 2
        df['fare_per_mile'] = np.where(
            df['trip_distance'] > 0,
            df['fare_amount'] / df['trip_distance'],
            np.nan
        )
        
        # Calculate idle time estimate - Feature 3
        # Assume average city speed is 12 mph, any time beyond that is idle
        df['idle_time_minutes'] = np.where(
            df['trip_distance'] > 0,
            np.maximum(0, (df['trip_duration'] / 60) - (df['trip_distance'] / 12 * 60)),
            np.nan
        )
        
        logger.info("Derived features calculated:")
        logger.info(f"- Trip speed: {df['trip_speed_mph'].notna().sum()} valid calculations")
        logger.info(f"- Fare per mile: {df['fare_per_mile'].notna().sum()} valid calculations")
        logger.info(f"- Idle time: {df['idle_time_minutes'].notna().sum()} valid calculations")
        
        return df
    
    def _remove_outliers(self, df: pd.DataFrame) -> pd.DataFrame:
        """Remove statistical outliers"""
        logger.info("Removing outliers")
        
        initial_count = len(df)
        
        # Remove trips with unrealistic duration (< 30 seconds or > 24 hours)
        duration_outliers = (df['trip_duration'] < 30) | (df['trip_duration'] > 86400)
        duration_outlier_count = duration_outliers.sum()
        
        if duration_outlier_count > 0:
            self.excluded_records.append({
                'reason': 'duration_outliers',
                'count': duration_outlier_count
            })
            logger.info(f"Excluding {duration_outlier_count} duration outliers")
        
        df = df[~duration_outliers]
        
        # Remove trips with unrealistic speed (> 100 mph or < 0.1 mph for moving trips)
        speed_outliers = (
            (df['trip_speed_mph'] > 100) | 
            ((df['trip_speed_mph'] < 0.1) & (df['trip_distance'] > 0.1))
        )
        speed_outlier_count = speed_outliers.sum()
        
        if speed_outlier_count > 0:
            self.excluded_records.append({
                'reason': 'speed_outliers',
                'count': speed_outlier_count
            })
            logger.info(f"Excluding {speed_outlier_count} speed outliers")
        
        df = df[~speed_outliers]
        
        # Remove trips with unrealistic fare per mile (> $50/mile)
        fare_per_mile_outliers = df['fare_per_mile'] > 50
        fare_outlier_count = fare_per_mile_outliers.sum()
        
        if fare_outlier_count > 0:
            self.excluded_records.append({
                'reason': 'fare_per_mile_outliers',
                'count': fare_outlier_count
            })
            logger.info(f"Excluding {fare_outlier_count} fare per mile outliers")
        
        df = df[~fare_per_mile_outliers]
        
        final_count = len(df)
        outlier_count = initial_count - final_count
        
        logger.info(f"Outlier removal complete. Excluded {outlier_count} outliers")
        
        return df
    
    def get_cleaning_summary(self) -> Dict:
        """Get summary of data cleaning process"""
        total_excluded = sum(record['count'] for record in self.excluded_records)
        total_suspicious = sum(record['count'] for record in self.suspicious_records)
        
        return {
            'excluded_records': self.excluded_records,
            'suspicious_records': self.suspicious_records,
            'total_excluded': total_excluded,
            'total_suspicious': total_suspicious
        }
    
    def save_cleaning_log(self, output_path: str):
        """Save cleaning log to file"""
        import json
        
        log_data = {
            'timestamp': datetime.now().isoformat(),
            'summary': self.get_cleaning_summary()
        }
        
        with open(output_path, 'w') as f:
            json.dump(log_data, f, indent=2)
        
        logger.info(f"Cleaning log saved to {output_path}")


